### ISSUE TITLE
Zeitwerk is 10x slower while eager loading empty directories on ruby-3.1 

### ISSUE THREAD

**Issue by ojab** [fxn/zeitwerk#200]
Pretty sure it's ruby issue, but I dunno how to craft a reproducer, so plz halp.
 
In the test repo https://github.com/ojab/tst, new rails app with 
```shell
for I in {a..z}; do for J in {a..z}; do mkdir -p app/foo/${J}/${I}; done; done
for I in {a..z}; do for J in {a..z}; do touch app/foo/${J}/${I}/.keep; done; done
```
locally there is:
```
$ ruby -v; time bundle exec rails r 'puts'
ruby 3.0.3p157 (2021-11-24 revision 3fb7d2cadc) [x86_64-linux]

real	0m6,569s
user	0m5,947s
sys	0m0,445s
```
and
```
$ ruby -v; time bundle exec rails r 'puts'
ruby 3.1.0p0 (2021-12-25 revision fb4df44d16) [x86_64-linux]

real	0m30,425s
user	0m11,838s
sys	0m17,763s
```
Each `const_get` in `Zeitwerk::Loader::Helpers#cget` jumped from around `0.006` sec to around `0.04` sec.

Found via application with frontend code in `app/front` by sudden spike in deployment time (30 sec -> 2 min) after update to ruby-3.1.

**Comment by fxn** [comment]
WOW. I could reproduce without Rails:

```ruby
require 'fileutils'
require 'tmpdir'
require 'zeitwerk'

Dir.mktmpdir do |dir|
  Dir.chdir(dir)

  for i in 'a'..'z'
    for j in 'a'..'z'
      FileUtils.mkdir_p("lib/#{i}/#{j}")
    end
  end

  loader = Zeitwerk::Loader.new
  loader.push_dir('lib')
  loader.setup

  start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  loader.eager_load
  puts Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
end
```

Interestingly, if we only use files, it takes slightly more, but by no means as much as with the example above:

```ruby
require 'fileutils'
require 'tmpdir'
require 'zeitwerk'

Dir.mktmpdir do |dir|
  Dir.chdir(dir)

  FileUtils.mkdir('lib')
  for i in 'a'..'z'
    for j in 'a'..'z'
      File.write("lib/#{i}_#{j}.rb", "#{i.upcase}#{j.upcase} = 1")
    end
  end

  loader = Zeitwerk::Loader.new
  loader.push_dir('lib')
  loader.setup

  start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  loader.eager_load
  puts Process.clock_gettime(Process::CLOCK_MONOTONIC) - start
end
```

This does not seem to happen with Zeitwerk 2.5.1.

**Comment by fxn** [comment]
Definitely, the root cause is pushing a path to `$LOADED_FEATURES`

```ruby
$LOADED_FEATURES << path
```

when implicit directories are eager loaded ([here](https://github.com/fxn/zeitwerk/blob/1ff6336d937514c923f2c5a60cfaa501f150a7b5/lib/zeitwerk/kernel.rb#L32)).

For some reason, in Ruby 3.1 that push is orders of magnitude slower than in Ruby 3.0.

@ojab while this is investigated, you could depend on 2.5.1 (which does not perform that push).

**Comment by fxn** [comment]
I have been able to see which is the commit that is responsible for the majority of the regression, but still not all of it.

I have reported the regression in https://bugs.ruby-lang.org/issues/18452.

**Comment by ngan** [comment]
Just another datapoint...we're seeing performance regression on Zeitwerk 2.5.2, Ruby 2.7.5, and Rails 6.1.4.4.

**Comment by ojab** [comment]
This has different root cause, this particular issue is only about regression in ruby-3.1, not in zeitwerk itself.
Maybe it would lead to a change in zeitwerk that would offset some performance for other cases, but better to file another issue with different reproducer.

**Comment by fxn** [comment]
@ngan 2.5.2 may be slightly slower for implicit namespaces because it does more things. Can you quantify the regression in a different ticket please?

**Comment by fxn** [comment]
The conclusion of the discussion in Redmine is that the change in 2.5.2 should be reverted. I have published 2.5.3 with the revert.

@ngan that should restore performance in your app too.

**Comment by ngan** [comment]
Awesome. Great work as always, @fxn !
